File: xo.kt
// Build
// kotlinc main.kt -include-runtime -d main.jar
// Run
// java -jar main.jar

val EMPTY_CHAR = '_'
val GAME_TIE = 0
val GAME_OVER = -1
val GAME_CONTINUE = -2
fun main()
{
    val myBoard = Board(EMPTY_CHAR)
    val playerOne = Player(1, 'X')
    val playerTwo = Player(2, 'O')
    
    var currentPlayer = playerOne
    while (true)
    {
        myBoard.print()
        val turnInt = currentPlayer.id
        println("Player $turnInt, enter a digits between 0 and 2")

        /* Currently - get two random numbers without override check */
        val (a, b) = getInts()
        myBoard.update(a, b, currentPlayer.symbol)
        val state = myBoard.gameState()

        if (state == GAME_CONTINUE) 
        {
            currentPlayer = if (currentPlayer == playerOne) playerTwo else playerOne
            continue
        }
        
        myBoard.print()
        if (state == GAME_TIE)
        {
            println("It's a tie!")
            break;
        }

        println("Player $turnInt won!")
        break;
    }
}

/* Reads two integers from input */
fun readInts() = readLine()!!.split(' ').map { it.toInt() }

/* Returns two random numbers */
fun getInts() = arrayOf((0..2).random(), (0..2).random())

class Player
{
    val id : Int
    val symbol : Char
    constructor(id_ : Int, symbol_ : Char)
    {
        id = id_
        symbol = symbol_
    }
}

class Board
{
    val b: Array< Array<Char> >
    constructor(emptyChar : Char)
    {
        b = Array(3) { Array(3) {emptyChar}}
    }

    fun print()
    {
        b.forEach{ println(it.contentToString()) }
    }

    fun update(i : Int, j : Int, sym : Char)
    {
        if (i < 0 || i > 2 || j < 0 || j > 2) return
        b[i][j] = sym
    }

    fun gameState() : Int
    {
        var emptyCharIn = false
        var over : Boolean
        
        // diagonals
        if (b[0][0] == b[1][1] && b[1][1] == b[2][2] && b[2][2] != EMPTY_CHAR) return GAME_OVER
        if (b[0][2] == b[1][1] && b[1][1] == b[2][0] && b[2][0] != EMPTY_CHAR) return GAME_OVER

        for (i in 0..2)
        {
            over = b[i][0] != EMPTY_CHAR
            over = over && b[i][0] == b[i][1] 
            over = over && b[i][1] == b[i][2]
            if  (over) return GAME_OVER

            over = b[0][i] != EMPTY_CHAR
            over = over && b[0][i] == b[1][i] 
            over = over && b[1][i] == b[2][i]
            if  (over) return GAME_OVER
        
            emptyCharIn = emptyCharIn || (EMPTY_CHAR in b[i])
        }
        
        if (!emptyCharIn) return GAME_TIE
        return GAME_CONTINUE
    }
}

File: xo.cs
const char space = '\t';
const char EC = '-'; // empty char
char[, ] board = {
                {EC, EC, EC},
                {EC, EC, EC},
                {EC, EC, EC}
                };
const int gameCONTINUE = -1;
const int gameTIE = 0;
const int gameP1 = 1;
const int gameP2 = 2;

bool notOver = true;
int turn = gameP1;
while (notOver)
{
    Console.Clear();
    printMatrix(board);
    Console.WriteLine("Player {0}, enter two digits from 0 to 2", turn);
    bool valid = updatePosition(board, turn); // Get (i, j) from user input and update
    

    if (!valid) continue;
    int winner = getWinner(board, turn);
    if (winner == gameCONTINUE)
    {
        if (turn == gameP1) turn = gameP2; else turn = gameP1;
        continue; // No winner, continue
    }
    
    printMatrix(board);
    if (winner != gameTIE)
        Console.WriteLine("Player {0} wins!", winner);
    else
        Console.WriteLine("It's a tie!");
    notOver = false;

    break;
}


static int getWinner(char[, ] b, int t)
{
    // Diagonal
    bool same = b[0, 0] == b[1, 1];
    same = same && b[1, 1] == b[2, 2]; 
    same = same && b[0, 0] != EC;
    if (same) return t;

    for (int i = 0; i < 3; i ++)
    {
        // Vertical
        same = b[i, 0] == b[i, 1];
        same = same && b[i, 1] == b[i, 2]; 
        same = same && b[i, 0] != EC;

        if (same) return t;

        // Horizontal
        same = b[0, i] == b[1, i];
        same = same && b[1, i] == b[2, i]; 
        same = same && b[0, i] != EC;

        if (same) return t;
    }
    if (contains(b, EC)) return gameCONTINUE;
    return gameTIE;
}


static bool contains(char[, ] b, char ec)
{
    for (int i = 0; i < b.GetLength(0); i++)
    {
        for (int j = 0; j < b.GetLength(1); j++)
        {
            if (b[i, j] == ec) return true;
        }
    }
    return false;
}


static bool updatePosition(char[, ] b, int t)
{
    int i, j;
    try
    {
        String? input = Console.ReadLine();
        if (input == null) return false;
        String[] splitLine = input.Split();
        if (splitLine == null || splitLine.Length != 2) return false;
        i = int.Parse(splitLine[0]);
        j = int.Parse(splitLine[1]);

        if (b[i, j] != EC) return false;

        b[i, j] = turnToChar(t);
    }
    catch (Exception)
    {
        return false;
    }
    return true;
}


static char turnToChar(int t)
{
    if (t == 1) return 'X';
    return 'O';
}


static void printMatrix(char[, ] b)
{
    for (int i = 0; i < b.GetLength(0); i++)
    {
        for (int j = 0; j < b.GetLength(1); j++)
        {
            Console.Write("{0}{1}", b[i, j], space);
        }
        Console.WriteLine();
    }
}